README.md
[![Published on NPM](https://img.shields.io/npm/v/@api-components/api-type-document.svg)](https://www.npmjs.com/package/@api-components/api-type-document)

[![Build Status](https://travis-ci.com/advanced-rest-client/api-type-document.svg)](https://travis-ci.com/advanced-rest-client/api-type-document)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/advanced-rest-client/api-type-document)

## &lt;api-type-document&gt;

A documentation table for RAML type / OAS schema properties. Works with AMF data model.

## Version compatibility

This version only works with AMF model version 2 (AMF parser >= 4.0.0).
For compatibility with previous model version use `3.x.x` version of the component.

## Styling

`<api-type-document>` provides the following custom properties and mixins for styling:

Custom property | Description | Default
----------------|-------------|----------
`--api-type-document` | Mixin applied to this elment | `{}`
`--api-type-document-union-button-background-color` | Background color of union selector button | `#fff`
`--api-type-document-union-button-color` | Color of union selector button | `#000`
`--api-type-document-union-button-active-background-color` | Background color of active union selector button | `#CDDC39`
`--api-type-document-union-button-active-color` | Color of active union selector button | `#000`

From `property-shape-document`

Custom property | Description | Default
----------------|-------------|----------
`--property-shape-document` | Mixin applied each proeprty element | `{}`
`--property-shape-document-array-color` | Property border color when type is an array | `#8BC34A`
`--property-shape-document-object-color` | Property border color when type is an object | `#FF9800`
`--property-shape-document-union-color` | Property border color when type is an union | `#FFEB3B`
`--arc-font-subhead` | Theme mixin, applied to the property title | `{}`
`--property-shape-document-title` | Mixin applied to the property title | `{}`
`--api-type-document-property-parent-color` | Color of the parent property label | `#757575`
`--api-type-document-property-color` | Color of the property name label when display name is used | `#757575`
`--api-type-document-child-docs-margin-left` | Margin left of the item's properties description relative to the title when the item is a child property of another property | `24px`
`--api-type-document-type-color` | Color of the "type" trait | `white`
`--api-type-document-type-background-color` | Background color of the "type" trait | `#2196F3`
`--api-type-document-trait-background-color` | Background color to main range trait (type, required, enum) | `#EEEEEE`,
`--api-type-document-trait-border-radius` | Border radious of a main property traits like type, required, enum | `3px`

From `property-range-document`

Custom property | Description | Default
----------------|-------------|----------
`--property-range-document` | Mixin applied to this elment | `{}`
`--api-type-document-type-attribute-color` | Color of each attribute that describes a property | `#616161`
`--api-type-document-examples-title-color` | Color of examples section title | ``
`--api-type-document-examples-border-color` | Example section border color | `transparent`
`--code-background-color` | Background color of the examples section | ``
`--arc-font-body1` | Mixin applied to an example name label | `{}`
`--arc-font-body2` | Mixin applied to the examples section title | `{}`

## Usage

### Installation
```
npm install --save @api-components/api-type-document
```

### In an html file

```html
<html>
  <head>
    <script type="module">
      import '@api-components/api-type-document/api-type-document.js';
    </script>
  </head>
  <body>
    <api-type-document></api-type-document>
  </body>
</html>
```

### In a LitElement

```js
import { LitElement, html } from 'lit-element';
import '@api-components/api-type-document/api-type-document.js';

class SampleElement extends PolymerElement {
  render() {
    return html`
    <api-type-document .amf="${this.amf}"></api-type-document>
    `;
  }
}
customElements.define('sample-element', SampleElement);
```

## Development

```sh
git clone https://github.com/advanced-rest-client/api-type-document
cd api-type-document
npm install
```

### Running the demo locally

```sh
npm start
```

### Running the tests
```sh
npm test
```


CHANGELOG.md
<a name="2.0.19"></a>
## 2.0.19 (2018-11-27)


### Breaking

* Updating AMF model to latest version ([6089daaea7c69db5fa3fcf7110e56b27ad7f9d7a](https://github.com/advanced-rest-client/api-type-document/commit/6089daaea7c69db5fa3fcf7110e56b27ad7f9d7a))

### Docs

* Updated documentation. ([d087117a6ade5b2bc567baf915b1c945b43bb42e](https://github.com/advanced-rest-client/api-type-document/commit/d087117a6ade5b2bc567baf915b1c945b43bb42e))
* Updated documentation. ([3a58d290f9269db7391507445c48b112bfbd3b15](https://github.com/advanced-rest-client/api-type-document/commit/3a58d290f9269db7391507445c48b112bfbd3b15))
* Updating documentation ([b529a3a2b4e49a4b75dabba132bd13e98dd7b6f0](https://github.com/advanced-rest-client/api-type-document/commit/b529a3a2b4e49a4b75dabba132bd13e98dd7b6f0))
* Updating documentation ([52e3fa35bffcd763bfc382bb7d4986094e817a7c](https://github.com/advanced-rest-client/api-type-document/commit/52e3fa35bffcd763bfc382bb7d4986094e817a7c))
* Updating documentation ([db9b0db64b5e905e00b333fce98084c7952c296d](https://github.com/advanced-rest-client/api-type-document/commit/db9b0db64b5e905e00b333fce98084c7952c296d))
* Updating documentation ([e091c01b80a6ffa91cd0b8944718feb7318941ea](https://github.com/advanced-rest-client/api-type-document/commit/e091c01b80a6ffa91cd0b8944718feb7318941ea))
* Updating documentation ([7f4cb0c0403ad25b2be292775add90076dc0b368](https://github.com/advanced-rest-client/api-type-document/commit/7f4cb0c0403ad25b2be292775add90076dc0b368))

### Fix

* Adding array test ([c501b9f8245b5183a24cd63309477f8212141195](https://github.com/advanced-rest-client/api-type-document/commit/c501b9f8245b5183a24cd63309477f8212141195))
* Adding missing resolve function call on the shape decument. ([e88e2cd2e0886184dde9baed611814c12812dc86](https://github.com/advanced-rest-client/api-type-document/commit/e88e2cd2e0886184dde9baed611814c12812dc86))
* Fixed union type selection ([299a5af80dc5cb24cefa3838b531ce09e45c4008](https://github.com/advanced-rest-client/api-type-document/commit/299a5af80dc5cb24cefa3838b531ce09e45c4008))
* Fixing APIC-83 ([43509e8bfeb95e4beaa19262619daf4381757745](https://github.com/advanced-rest-client/api-type-document/commit/43509e8bfeb95e4beaa19262619daf4381757745))
* Fixing component after AMF model change. ([31d0f3fd822e810e73f8c555ada8ef62827d02c8](https://github.com/advanced-rest-client/api-type-document/commit/31d0f3fd822e810e73f8c555ada8ef62827d02c8))
* Fixing data types keys read ([462a513c2364f1a2b84e342942d568aaa774171e](https://github.com/advanced-rest-client/api-type-document/commit/462a513c2364f1a2b84e342942d568aaa774171e))
* Fixing hidden state read. ([017be03de08bab957bf4411dfeb1ad584df03b31](https://github.com/advanced-rest-client/api-type-document/commit/017be03de08bab957bf4411dfeb1ad584df03b31))
* Fixing model access properties names ([d1969f2e3c653392c406134e83568001b94cda7b](https://github.com/advanced-rest-client/api-type-document/commit/d1969f2e3c653392c406134e83568001b94cda7b))
* Fixing nil and any shape recognition ([ae5f6c62b4b55c94f59d34ca12d4012118ccc029](https://github.com/advanced-rest-client/api-type-document/commit/ae5f6c62b4b55c94f59d34ca12d4012118ccc029))
* Fixing styling issues ([eeec0046d022ef57bdaf7cc091ce9a5e2ef80843](https://github.com/advanced-rest-client/api-type-document/commit/eeec0046d022ef57bdaf7cc091ce9a5e2ef80843))
* Fixing test for new model ([50fe356f781ca1443f7965ee954e12391651b114](https://github.com/advanced-rest-client/api-type-document/commit/50fe356f781ca1443f7965ee954e12391651b114))
* Fixing types render after API change ([33d04bf4be9c905cd78accd034f09d30ef83b1e7](https://github.com/advanced-rest-client/api-type-document/commit/33d04bf4be9c905cd78accd034f09d30ef83b1e7))

### New

* Adding recursive data type ([a87aee284486c5cdcf159762115cfe7de74fa760](https://github.com/advanced-rest-client/api-type-document/commit/a87aee284486c5cdcf159762115cfe7de74fa760))
* Adding support for OAS's "and" type. ([5b5defba3e5e830022dd5497f358a67e2a82d7fe](https://github.com/advanced-rest-client/api-type-document/commit/5b5defba3e5e830022dd5497f358a67e2a82d7fe))

### Update

* Adding support for compact model ([648a4c879b1b6be413c0ce05ecee61cb1a491804](https://github.com/advanced-rest-client/api-type-document/commit/648a4c879b1b6be413c0ce05ecee61cb1a491804))
* Annotation element now can access full AMF data ([4e0f6e4a3bf7c20820d400cf21343bc786676437](https://github.com/advanced-rest-client/api-type-document/commit/4e0f6e4a3bf7c20820d400cf21343bc786676437))
* Fixing overflow issues with documentation. ([00b55afa6b3b57ee60713424f0af3a3956bcb952](https://github.com/advanced-rest-client/api-type-document/commit/00b55afa6b3b57ee60713424f0af3a3956bcb952))
* Sumplifying styling API and views ([99f55d651c8ea52c77e5138a0a51a8e3c1d53493](https://github.com/advanced-rest-client/api-type-document/commit/99f55d651c8ea52c77e5138a0a51a8e3c1d53493))
* Test model generator is not external module ([f2b1411b128cab09c82aa939f15ce094ea544c08](https://github.com/advanced-rest-client/api-type-document/commit/f2b1411b128cab09c82aa939f15ce094ea544c08))
* Updated Travis configuration to connect to Sauce Labs. ([217dd0945050383a70dd9118a3a6c9e2a19639cf](https://github.com/advanced-rest-client/api-type-document/commit/217dd0945050383a70dd9118a3a6c9e2a19639cf))
* Updating default styles for examples. ([5b85407cc138cc04a4dab8738c35c3fcd85473c1](https://github.com/advanced-rest-client/api-type-document/commit/5b85407cc138cc04a4dab8738c35c3fcd85473c1))
* updating default styles. ([7268574b90cbc29d2c1835b32ecaf42f0100def9](https://github.com/advanced-rest-client/api-type-document/commit/7268574b90cbc29d2c1835b32ecaf42f0100def9))
* Updating label rendering for property name and display name ([072f943ae8b918bc842f0b296031f98d51cc8a33](https://github.com/advanced-rest-client/api-type-document/commit/072f943ae8b918bc842f0b296031f98d51cc8a33))

src/ShapeStyles.js
import { css } from 'lit-element';

export default css`
  :host {
    display: block;
    border-bottom-width: 1px;
    border-bottom-color: var(
      --property-shape-document-border-bottom-color,
      #cfd8dc
    );
    border-bottom-style: var(
      --property-shape-document-border-bottom-style,
      dashed
    );
    padding: var(--property-shape-document-padding);
  }

  :host(:last-of-type) {
    border-bottom: none;
  }

  [hidden] {
    display: none !important;
  }

  .property-title {
    font-size: var(
      --property-shape-document-title-font-size,
      var(--arc-font-subhead-font-size)
    );
    font-weight: var(
      --property-shape-document-title-font-weight,
      var(--arc-font-subhead-font-weight)
    );
    line-height: var(
      --property-shape-document-title-line-height,
      var(--arc-font-subhead-line-height)
    );

    margin: 4px 0 4px 0;
    font-size: 1rem;
    font-weight: var(--api-type-document-property-title-font-weight, 500);
    word-break: break-word;
    color: var(--api-type-document-property-title-color);
  }

  .property-title[secondary] {
    font-weight: var(
      --api-type-document-property-title-secondary-font-weight,
      400
    );
    color: var(--api-type-document-property-title-secondary-color, #616161);
  }

  .parent-label {
    color: var(--api-type-document-property-parent-color, #757575);
  }

  .property-display-name {
    font-weight: var(--api-type-document-property-name-font-weight, 500);
    color: var(
      --api-type-document-property-name-color,
      var(--api-type-document-property-color, #212121)
    );
    margin: 4px 0 4px 0;
    font-size: var(--api-type-document-property-name-font-size, 16px);
  }

  api-type-document,
  property-range-document {
    transition: background-color 0.4s linear;
  }

  :host([isobject]) api-type-document.complex,
  :host([isunion]) api-type-document.complex,
  :host([isarray]) api-type-document.complex {
    padding-left: var(--api-type-document-child-docs-padding-left, 20px);
    margin-left: var(--api-type-document-child-docs-margin-left, 0px);
    padding-right: var(--api-type-document-child-docs-padding-right, initial);
  }

  :host([isobject]) api-type-document,
  :host([isobject]) property-range-document {
    border-left-color: var(--property-shape-document-object-color, #ff9800);
    border-left-width: 2px;
    border-left-style: solid;
    padding-left: 12px;
  }

  :host([isarray]) api-type-document,
  :host([isarray]):not([isscalararray]) property-range-document {
    border-left: 2px var(--property-shape-document-array-color, #8bc34a) solid;
    padding-left: 12px;
  }

  :host([isunion]) api-type-document,
  :host([isunion]) property-range-document {
    border-left: 2px var(--property-shape-document-union-color, #ffeb3b) solid;
    padding-left: 12px;
  }

  .property-traits {
    display: -ms-flexbox;
    display: flex;
    flex-direction: row;
    -ms-flex-direction: row;
    flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    margin-bottom: 8px;
  }

  .property-traits > span {
    display: inline-block;
    margin-right: 8px;
    padding: var(--api-type-document-trait-padding, 2px 4px);
    background-color: var(--api-type-document-trait-background-color, #eeeeee);
    color: var(--api-type-document-trait-color, #616161);
    border-radius: var(--api-type-document-trait-border-radius, 3px);
    font-size: var(--api-type-document-trait-font-size, 13px);
  }

  .property-traits > span.data-type {
    background-color: var(--api-type-document-type-background-color, #1473bf);
    color: var(--api-type-document-type-color, white);
    padding: var(--api-type-document-trait-data-type-padding, 2px 4px);
    font-weight: var(--api-type-document-trait-data-type-font-weight, normal);
  }

  arc-marked {
    background-color: transparent;
    padding: 0;
    margin-top: 20px;
    margin-bottom: 20px;
  }

  .link-label {
    text-decoration: underline;
    cursor: pointer;
  }

  .shape-header {
    display: -ms-flexbox;
    display: flex;
    -ms-flex-direction: row;
    flex-direction: row;
    -ms-flex-align: center;
    align-items: start;
  }

  .name-area {
    -ms-flex: 1;
    flex: 1;
  }
`;
